gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/policy/impl/ConditionalReadWriteStream.java

```diff
@@ -0,0 +1,179 @@
+    public SimpleReadWriteStream<Buffer> endHandler(Handler<Void> handler) {
+        delegate.endHandler(handler);
+        return super.endHandler(handler);
+    }
+
+    /**
+     * Pause both delegate and this instance in case of full queue
+     * @return this instance of ReadStream<Buffer>
+     */
+    @Override
+    public ReadStream<Buffer> pause() {
+        delegate.pause();
+        return super.pause();
+    }
+
+    /**
+     * Resume both delegate and this instance in case of full queue
+     * @return this instance of ReadStream<Buffer>
+     */
+    @Override
+    public ReadStream<Buffer> resume() {
+        delegate.resume();
+        return super.resume();
+    }
+
+    /**
+     * When condition evaluate to true, then we write on the delegate (targeted policy), else, on this instance
+     * @param buffer is the one to write
+     * @return this instance to be able to continue the chaining
+     */
+    @Override
+    public SimpleReadWriteStream<Buffer> write(Buffer buffer) {
+        if (evaluateCondition()) {
+            delegate.write(buffer);
+        } else {
+            super.write(buffer);
+        }
+        return this;
+    }
+
+    /**
+     * When condition evaluate to true, we end the delegate (targeted policy), else, we just end the current instance
+     */
+    @Override
+    public void end() {
+        if (evaluateCondition()) {
+            delegate.end();
+        } else {
+            super.end();
+        }
+    }
+
+    /**
+     * When condition evaluate to true, we end the delegate (targeted policy), else, we just end the current instance
+     */
+    @Override
+    public void end(Buffer buffer) {
+        if (evaluateCondition()) {
+            delegate.end(buffer);
+        } else {
+            super.end(buffer);
+        }
+    }
+
+    /**
+     * Drain handler for both delegate (targeted policy) and this instance.
+     * @param drainHandler to release the pressure on the stream
+     * @return this instance as a WriteStream<Buffer>
+     */
+    @Override
+    public WriteStream<Buffer> drainHandler(Handler<Void> drainHandler) {
+        delegate.drainHandler(drainHandler);
+        return super.drainHandler(drainHandler);
+    }
+
+    /**
+     * Depending on condition evaluation, get writeQueueFull()
+     * @return true if write queue is full
+     */
+    @Override
+    public boolean writeQueueFull() {
+        return evaluateCondition() ? delegate.writeQueueFull() : super.writeQueueFull();
+    }
+
+    /**
+     * Evaluate condition once and keep its value for this instance. This avoid evaluating multiple times the condition.
+     * In case of Exception during evaluation, policy chain will fail and exception rethrown
+     * @return the evaluation result of the condition.
+     */
+    private Boolean evaluateCondition() {
+        if (isConditionTruthy == null) {
+            try {
+                isConditionTruthy = conditionEvaluator.evaluate(context, condition);
```

Comment:
Ok, forget my first comment, just see that you manage single condition evaluation here.